File: apps/api/app/schemas/vendor.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.order import FulfillmentStatus, OrderStatus, PaymentMethod


class FulfillmentStatusUpdate(BaseModel):
    status: FulfillmentStatus
    courier_tracking: str | None = Field(default=None, max_length=255)


class VendorFulfillmentRead(BaseModel):
    """Fulfillment enriched with order context — used in the vendor portal."""
    id: uuid.UUID
    order_id: uuid.UUID
    public_token: uuid.UUID
    order_status: OrderStatus
    payment_method: PaymentMethod | None
    status: FulfillmentStatus
    delivery_date: date
    delivery_slot: str | None
    recipient_name: str
    recipient_phone: str
    address_line1: str
    address_line2: str | None
    city_name: str
    landmark: str | None
    courier_tracking: str | None
    dispatched_at: datetime | None
    delivered_at: datetime | None
    item_count: int
    subtotal_pkr: int

File: apps/web/lib/vendor.ts
import { apiFetch } from "@/lib/api";

export interface VendorRead {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  city_id: string;
  is_active: boolean;
}

export interface VendorUpdate {
  name?: string;
  description?: string | null;
  logo_url?: string | null;
}

export type FulfillmentStatus =
  | "pending"
  | "preparing"
  | "ready"
  | "dispatched"
  | "out_for_delivery"
  | "delivered"
  | "failed";

export interface VendorFulfillmentRead {
  id: string;
  order_id: string;
  public_token: string;
  order_status: string;
  payment_method: string | null;
  status: FulfillmentStatus;
  delivery_date: string;
  delivery_slot: string | null;
  recipient_name: string;
  recipient_phone: string;
  address_line1: string;
  address_line2: string | null;
  city_name: string;
  landmark: string | null;
  courier_tracking: string | null;
  dispatched_at: string | null;
  delivered_at: string | null;
  item_count: number;
  subtotal_pkr: number;
}

export interface FulfillmentStatusUpdate {
  status: FulfillmentStatus;
  courier_tracking?: string | null;
}

export interface VendorProductRead {
  id: string;
  name: string;
  slug: string;
  base_price_pkr: number;
  is_active: boolean;
}

export const vendorApi = {
  getMe: (token: string) =>
    apiFetch<VendorRead>("/vendor/me", { token }),

  updateMe: (body: VendorUpdate, token: string) =>
    apiFetch<VendorRead>("/vendor/me", { method: "PATCH", body, token }),

  listFulfillments: (token: string, filterStatus?: FulfillmentStatus) => {
    const qs = filterStatus ? `?filter_status=${filterStatus}` : "";
    return apiFetch<VendorFulfillmentRead[]>(`/vendor/fulfillments${qs}`, { token });
  },

  updateFulfillment: (id: string, body: FulfillmentStatusUpdate, token: string) =>
    apiFetch<VendorFulfillmentRead>(`/vendor/fulfillments/${id}`, {
      method: "PATCH",
      body,
      token,
    }),

  listProducts: (token: string, page = 1) =>
    apiFetch<VendorProductRead[]>(`/vendor/products?page=${page}`, { token }),
};
